frontend/src/metabase-lib/Question.ts

```diff
@@ -313,14 +313,27 @@ class QuestionInner {
     return this._card && this._card.displayIsLocked;
   }
 
-  maybeResetDisplay(sensibleDisplays, previousSensibleDisplays): Question {
+  maybeResetDisplay(
+    data,
+    sensibleDisplays,
+    previousSensibleDisplays,
+  ): Question {
     const wasSensible =
       previousSensibleDisplays == null ||
       previousSensibleDisplays.includes(this.display());
     const isSensible = sensibleDisplays.includes(this.display());
     const shouldUnlock = wasSensible && !isSensible;
     const defaultDisplay = this.setDefaultDisplay().display();
 
+    // For 1x1 data, we show a scalar. If our display was a 1x1 type, but the data
+    // isn't 1x1, we show a table.
+    const isScalar = ["scalar", "progress", "gauge"].includes(this.display());
+    const isOneByOne = data.rows.length === 1 && data.cols.length === 1;
+    // if we have a 1x1 data result then this should always be viewed as a scalar
+    if (!isScalar && isOneByOne && !this.displayIsLocked()) {
+      return this.setDisplay("scalar");
+    }
```

Comment:
> This simply merges the `switchTableScalar` method into the `maybeResetDisplay` method

I don't think this is a good idea - `maybeResetDisplay` is already difficult to reason about (BTW the `maybe` prefix in the name is a good complexity indicator).

If anything, we should be breaking things down - if there is a need for a function that does so many things, it should be composed of smaller functions, that can be easily reasoned about.

A simple improvement would be to leave `switchTableScalar` as it was (we can make it `private` though), and just call it inside `maybeResetDisplay`.
But this comes at a price of adding 3rd argument to `maybeResetDisplay`, which I'm not a fan of.
That's why I'm leaning towards leaving things as they were.

----

How exactly would this change help with #31891?

Reply:
> How exactly would this change help with https://github.com/metabase/metabase/pull/31891?

This is an attempt to make auto visualization selection easier to understand, so I could fix [this bug](https://github.com/metabase/metabase/issues/33208), which was preventing me from getting the tests on that PR to pass. I have since fixed the bug [here]([Stop native parameters defaults changing chart type after save or load](https://github.com/metabase/metabase/pull/33211#top)) though, so actually this isn't needed anymore.

However, I would still like to land this refactoring. 

> `maybeResetDisplay` is already difficult to reason about

Making `maybeResetDisplay` easier to understand is not my goal. My goal is to make viz auto-selection easier to understand in its entirety. Currently I'd argue to understand viz auto-selection you need to understand the internals of how both `switchTableScalar` and `maybeResetDisplay` work, and how they interact. Having them be separate functions makes understanding harder, not easier imo. In fact, after I understood the internals of how these work I discovered two redundant conditional branches in `switchTableScalar` and removed them. You'll see the resulting code is actually shorter and easier to understand.

Reply:
> Having them be separate functions makes understanding harder, not easier imo. 

I disagree. Let's see what the rest of the team thinks.

The only problem I see with this is that the functions don't look related.
But if we make `maybeResetDisplay` use `switchTableScale` (like suggested in my [previous comment](https://github.com/metabase/metabase/pull/33185#discussion_r1294263155)), that problem goes away. 

----


> In fact, after I understood the internals of how these work I discovered two redundant conditional branches in switchTableScalar and removed them. You'll see the resulting code is actually shorter and easier to understand.

This is great but it's not related to merging the two functions.

Reply:
> This is great but it's not related to merging the two functions.

It's a direct result from understanding the internals of both the functions together, which I find easier when their bodies are merged into one function because their conditional branches can be made exclusive of each other.

I've extracted out `switchTableScalar` and called it from `maybeResetDisplay` as you suggested anyways. The difference doesn't matter so much to me.